llunveil/llunveil.c

```diff
@@ -124,6 +138,27 @@ static int llunveil_init()
         }
         return -1;
     }
+
+    ruleset_attr.handled_access_fs = ACCESS_FS_READ | ACCESS_FS_WRITE
+       | ACCESS_FS_CREATE | ACCESS_FS_EXECUTE;
+    // Limit the available ruleset attributes to the kernel's supported ones.
+    // WORKARROUND: Apprantly `handled_access_fs` as shown on the documentation
+    // does not work on the kernel shipped with WSL2. It just spits out erorr.
+    // Also having trouble with Ubuntu 22.04
+    if(abi_version < 3) {
+        ruleset_attr.handled_access_fs &= ~LANDLOCK_ACCESS_FS_TRUNCATE;
+        ruleset_mask &= ~LANDLOCK_ACCESS_FS_TRUNCATE;
+    }
+    if(abi_version < 2) {
+        ruleset_attr.handled_access_fs &= ~LANDLOCK_ACCESS_FS_REFER;
+        ruleset_mask &= ~LANDLOCK_ACCESS_FS_REFER;
+    }
+
+    ruleset_fd = landlock_create_ruleset(&ruleset_attr, sizeof(ruleset_attr), 0);
+    if (ruleset_fd < 0) {
+        perror("Failed to create landlock ruleset");
+        return -1;
+    }
```

Comment:
@l0kod I followed the [landlock documentation](https://docs.kernel.org/userspace-api/landlock.html) to set `ruleset_attr.handled_access_fs`. But it tigers `EINVAL` upon syscall. I think I misunderstood the document or did something really stupid.. Am I?

The output is generated on Ubuntu 22.04 (kernel 5.19)

```
$ ./lljail -r /usr -rx /usr/lib -rx /usr/lib64 -rx /lib64 -rx /bin -rwc /tmp -r /etc -r $HOME -rw /dev -- /bin/bash
failed to update ruleset: Invalid argument
Could not populate ruleset for /tmp: Invalid argument
unveil() failed. Error: Invalid argumentfailed to update ruleset: Invalid argument
Could not populate ruleset for /dev: Invalid argument
unveil() failed. Error: Invalid argumentbash: /dev/null: Permission denied
bash: /dev/null: Permission denied
```

Reply:
Just from looking at the code, I wonder why the initial bit combination is made up without `LANDLOCK_` prefixes but the &= has. If that is a mistake, then also `ACCESS_FS_TRUNCATE` is missing in the initial list:

``` c
 ruleset_attr.handled_access_fs = ACCESS_FS_READ | ACCESS_FS_WRITE
       | ACCESS_FS_CREATE | ACCESS_FS_EXECUTE | ACCESS_FS_TRUNCATE; // ** added ACCESS_FS_TRUNCATE 
    if(abi_version < 3) {
        ruleset_attr.handled_access_fs &= ~ACCESS_FS_TRUNCATE; // **  removed LANDLOCK_
        ...
```
But I am just guessing....

Reply:
> @l0kod I followed the [landlock documentation](https://docs.kernel.org/userspace-api/landlock.html) to set `ruleset_attr.handled_access_fs`. But the `landlock_add_rule` still tigers `EINVAL` upon syscall if unsupported parameters are passed. I think I misunderstood the document or did something really stupid.. Am I?

You did right for the ruleset creation (i.e. masking `ruleset_attr.handled_access_fs`), but you need to do the same when adding a rule. You can remove the `ruleset_mask` variable and use `ruleset_attr.handled_access_fs` instead. The `-EINVAL` error means that you are trying to add a rule which [is not a subset of the ruleset handled access rights](https://docs.kernel.org/userspace-api/landlock.html#extending-a-ruleset). The documentation doesn't fully show this use case, there is only three static access rights, but I'll extend the documentation for clarification. You can also take a look at the [sandboxer example](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/samples/landlock/sandboxer.c).

Reply:
> Just from looking at the code, I wonder why the initial bit combination is made up without `LANDLOCK_` prefixes but the `&=` uses such a value. If that is a mistake, then also `ACCESS_FS_TRUNCATE` seems to be missing from the initial list:

The `ACCESS_*` bitmasks are specific to this unveil application and a subset of the kernel's `LANDLOCK_ACCESS_*` bitmasks. The removed bits for ABI < 3 are also removed for ABI <2.

Reply:
Thanks a lot! I thought `ruleset_attr.handled_access_fs` is a bit mask that the kernel will automatically apply upon calling `landlock_add_rule`. 

I feel the wording in the document is a bit confusing. It says we should check the ABI version and deny flags. Since the attribute is passed to `landlock_create_ruleset`. I assumed it's automatically applied for me since the kernel have the information. Glad to know I'm wrong and I need to read the docs more closely.
> Let’s check if we should remove the LANDLOCK_ACCESS_FS_REFER or LANDLOCK_ACCESS_FS_TRUNCATE access rights, which are only supported starting with the second and third version of the ABI.